databricks_cli/pipelines/api.py

```diff
@@ -38,28 +38,80 @@
 from databricks_cli.utils import InvalidConfigurationError
 
 BUFFER_SIZE = 1024 * 64
+base_pipelines_dir = 'dbfs:/pipelines/code/'
 
 
 class PipelinesApi(object):
     def __init__(self, api_client):
         self.client = DeltaPipelinesService(api_client)
         self.dbfs_client = DbfsApi(api_client)
 
+    def deploy(self, spec, headers=None):
+        """
+        Pipelines Deploy API
+        :param spec:
+        :param headers:
+        :return:
+        """
+        lib_objects = LibraryObject.convert_from_libraries(spec.get('libraries', []))
+        local_lib_objects, rest_lib_objects = \
+            self._partition_libraries_and_extract_local_paths(lib_objects)
+        remote_lib_objects = list(map(lambda llo:
+                                      LibraryObject(llo.lib_type, self._get_hashed_path(llo.path)),
+                                      local_lib_objects))
+        upload_files = self._get_files_to_upload(local_lib_objects, remote_lib_objects)
+
+        for llo, rlo in upload_files:
+            try:
+                self.dbfs_client.put_file(llo.path, rlo.path, False)
+            except Exception as e:
+                raise RuntimeError('Error \'{}\' while uploading {}'.format(e, llo.path))
+
+        rest_lib_objects.extend(remote_lib_objects)
+        spec['libraries'] = LibraryObject.convert_to_libraries(rest_lib_objects)
+        spec['credentials'] = self._get_credentials_for_request()
+        self.client.client.perform_query('PUT',
+                                         '/pipelines/{}'.format(spec['id']),
+                                         data=spec,
+                                         headers=headers)
+
+    def delete(self, pipeline_id, headers=None):
+        """
+        Pipelines Delete API
+        :param pipeline_id:
+        :param headers:
+        :return:
+        """
+        self.client.delete(pipeline_id, self._get_credentials_for_request(), headers)
+
     @staticmethod
     def _partition_libraries_and_extract_local_paths(lib_objects):
+        """
+        Partitions the given set of libraries into local and remote by checking uri scheme
+        :param lib_objects: List[LibraryObject]
+        :return: List[List[LibraryObject], List[LibraryObject]]
+        """
         local_lib_objects, rest_lib_objects = [], []
         for lib_object in lib_objects:
             uri_scheme = urllib.parse.urlsplit(lib_object.path).scheme
             if lib_object.lib_type == 'jar' and uri_scheme == '':
                 local_lib_objects.append(lib_object)
             elif lib_object.lib_type == 'jar' and uri_scheme.lower() == 'file':
+                if lib_object.path.startswith('file:////'):
+                    raise RuntimeError('Invalid file uri scheme')
                 local_lib_objects.append(LibraryObject(lib_object.lib_type, lib_object.path[5:]))
```

Comment:
1. https://en.wikipedia.org/wiki/File_URI_scheme#How_many_slashes? says that file:// is technically invalid. I'm still down to support it, but if you want to disallow it that's reasonable too.

2. That startswith check needs to be case insensitive.

3. If we strip off the first 5 characters, and the file starts with file:///path, we'll end up with ///path. Does this work when you try to read the file later on?

Reply:
1. It is technically invalid, but seems like it is still common. Another option is to be like dbfs, which only allows a single slash and nothing else. 

3. yes, ///path and //path and /path are equivalent when trying to open the file.